### ISSUE TITLE
Wrapping inserts a space which causes copy-paste and url clicking to fail

### ISSUE THREAD

**Issue by joshka** [Canop/bacon#125]
When the terminal width is not as wide as the message, the message wraps (which is good). The problem however is that at the wrap point, a space is inserted. This prevents terminal apps (iterm2 in my case) from detecting that a url has continued onto the next line. It also inserts a space into any text copied, which makes it annoying to copy past a suggested fix.

Example:
```
 1  warning: use Option::map_or_else instead of an if let/else
    --> src/view/login.rs:180:21
     |
 180 | /                     match self.authentication_code.as_ref() {
 181 | |                         Some(code) => Transition::To(LoginState::Com.
 ..
 182 | |                         None => Transition::None,
 183 | |                     }
     | |_____________________^ help: try: `self.authentication_code.as_ref().m
 ap_or_else(|| Transition::None, |code| Transition::To(LoginState::CompleteAut
 hentication(registered.clone(), code.clone())))`
     |
     = help: for further information visit https://rust-lang.github.io/rust-cl
 ippy/master/index.html#option_if_let_else
     = note: `-W clippy::option-if-let-else` implied by `-W clippy::nursery`
```

Causes two problems:
1. This cannot be pasted as a solution to the problem:
```
self.authentication_code.as_ref().m
 ap_or_else(|| Transition::None, |code| Transition::To(LoginState::CompleteAut
 hentication(registered.clone(), code.clone())))
```
2. This cannot be clicked to understand the problem better:
```
https://rust-lang.github.io/rust-cl
 ippy/master/index.html#option_if_let_else
```

Suggested fix:
Don't add the extra space when wrapping

**Comment by Canop** [comment]
If we remove this indentation, you'll still have the problem due to the scrollbar.

**Comment by joshka** [comment]
Hrm, I didn't realize that there was a scrollbar :D
You're right that is a problem. A workaround that could work might be to wrap urls in [OSC 8 sequences](https://gist.github.com/egmontkob/eb114294efbcd5adb1944c9f3cb5feda) (though perhaps that should be an upstream request on clippy. Do you happen to know how much processing of the clippy output gets done? Would OSC codes like this be passed through verbatim, or need to be explicitly handled by bacon?

TL;DR in case you've never seen this, the OSC 8 escape sequence:
```
printf '\e]8;;http://example.com/long-link-that-spans-multiple-lines\e\\https://example.com/long-link-\nthat-spans-multiple-lines\e]8;;\e\\\n'
```
Makes links clickable as the full link not the truncated link:
<img width="357" alt="image" src="https://user-images.githubusercontent.com/381361/229386032-6e795b29-a253-4b00-8587-d3af0be5cd04.png">


That addresse problem 2, but not problem 1. My guess is that problem 1 would only be fixable by turning off the scroll bar (perhaps moving the scroll position to the status bar?)

This is not a high priority item - more an annoyance than anything.

**Comment by Canop** [comment]
A possible approach:
- a click on a line visibly selects it
- a click on a title line visibly selects the whole item 
- a `copy` internal puts the selected text in the clipboard
- a `copy-unstyled` internal puts the selected text in the clipboard, without the TTY decorations

**Comment by tgrushka** [comment]
@Canop,

> If we remove this indentation, you'll still have the problem due to the scrollbar, as soon as the report doesn't fit the height.

I'm sorry, this is an accessibility barrier, and that argument is like the owner of a building refusing to add a wheelchair ramp because, "there are still stairs on the second floor." Just because one accessibility barrier might still exist, doesn't justify refusing to fix another major barrier that is even worse.

I am visually impaired and use
### `large fonts.`

Here's an example. In an app I'm building, I use `todo!("label")` in places I've yet to implement. And of course I have other errors, too. I want to go directly to the error location in my editor by control- or command-clicking the link in the terminal.

<img width="1884" alt="Image" src="https://github.com/user-attachments/assets/71049c6a-9c43-42d4-9d53-16b73312d584" />

Now I try to click the link and open it in my editor:
```
src/routes/auth_respon
 der.rs:110:34:
```

See? It wraps, with a space. Notice when I hover, only the first part of the link is highlighted, and instead of opening my editor, it opens my default browser to:
```
http://src/routes/auth_respon
```

Because the terminal only detects the link up to the space. Would anyone want that?

If the text wrapped like normal, like it should, like it does when running `cargo test`:

<img width="1902" alt="Image" src="https://github.com/user-attachments/assets/13f02a5e-f1cc-40a4-8d5f-230861021b0e" />

See? No extra space. Now I can control- / command-click the link:
```
src/routes/auth_responder.rs:110:34
```

And it opens directly in my editor. I don't have to strain my eyes to see what file it is, what line it's on, etc. I just click the link and go right to my error.

I think this would also help developers who are not visually impaired.

@joshka thank you for opening this issue.

**Comment by tgrushka** [comment]
Anyone else coming across this, my workaround is a wrapper script as follows:

(Maybe this "unwrapping" of links could be built in to bacon?)

bacon.toml:
```toml
[jobs.test]
# wrapper script to wrap links so bacon doesn't break them
command = ["./bacon-test.sh"]
# ... other options ...
```

The problem is, it has to be customised to handle the args you want and properly pass them to cargo.

bacon-test.sh (make executable):
```bash
#!/usr/bin/env bash

# This script is here to pre-wrap links for bacon so they remain clickable.

# Determine if a specific test was requested
TEST="$1"

# Set environment variables for better debugging
# export RUST_BACKTRACE=1
export RUST_LOG=debug

# Run the appropriate test command directly without capturing output
if [ -z "$TEST" ]; then
    # Run all tests
    OUTPUT=$(exec cargo test --all-features)
else
    # Run specific test
    OUTPUT=$(exec cargo test --test "$TEST" --all-features)
fi

# Without boldface links
# echo "$OUTPUT" | sed $'s| at |\\\n at |g' | sed $'/ at /s/$/\\\n/'

# With boldface links
echo "$OUTPUT" | sed $'s| at |\\\n at |g' | sed $'s| at \(.*:[0-9]*:[0-9]*\)| at \\\033[1m\\1\\\033[0m|g' | sed $'/ at /s/$/\\\n/'
```

Then output looks like this. Hooray, links are clickable again, and boldface!

<img width="1816" alt="Image" src="https://github.com/user-attachments/assets/9adb2b1f-6102-43ac-9021-d8d3d0ac2e79" />

**Comment by c-git** [comment]
@tgrushka I'm very sorry that this issue has created challenges for you. 

I would like to respond to your question which I have made bold in the quote below.

> Because the terminal only detects the link up to the space. **Would anyone want that?**

Without looking at the code I think it's very possible this is to prevent the text from showing up behind the scroll bar. The space is only added if the scroll bar is needed. If the text fits on one screen there is no scroll bar and no space is added. I do not think there was any desire to cause "clickable" text to be broken. It's very possible that this issue didn't have a trivial solution and there are only 24 hours in a day to allocate. Some users may not be affected if they use two screens as it might still fit on one line and as such the severity for some users may not be immediately obvious. The severity was not obvious to me at least and it doesn't open my browser in vscode so I didn't know that's how some editors would treat it.

**Comment by joshka** [comment]
@c-git reading between the lines here, it seems like you might have some imputed negative tone into @tgrushka's question and might be reacting a bit to that. It's difficult to convey tone in text, so I'd like to assume positive intent and suggest a more neutral reading that only looks at the words here. So perhaps "Is this something which is necessary", rather than "why would anyone be stupid enough to think this is something that would want" (I'm exaggerating a lot here ;)).

I agree with your logic on this however. There's a strong rational basis for why this happened. It's a fairly common accessibility problem in TUI apps which have non-end-of-line wrapping and the way that they interact with terminals which detect end of line soft wrapping in links. At its root this stems from that there's not a good connection between an TUI's ability to write characters to cells and for characters on multiple lines to be seen as one logical line.

I'm aware of only 1 TUI that solves this [lnav](https://lnav.org/), and which does so by having a shortcut key to display the currently visible lines without wrapping and scrollbars. This allows the terminal to interpret the wrapping correctly for links.

It's possible that using [OSC 8](https://gist.github.com/egmontkob/eb114294efbcd5adb1944c9f3cb5feda) for terminal hyperlinks might be a solution to this, but I'm not sure I know how file links should be handled. I'd guess that each terminal would likely handle this differently and this may make the situation worse.

@tgrushka the idea of formatting file location links onto a newline regardless of whether they would wrap could be worth considering pushing upstream into the rust compiler. It likely affects cargo, rustc, rustfmt, clippy, etc... The obvious downside is that this increases the vertical space used to report errors, so it's a loss when you consider information density for most users. But it's a clear win for accessibility. I don't know how to calculate that tradeoff. I imagine that this is a deeper problem to solve than this however, so I'm happy you have a solution that works here.

According to @estebank on a [https://lobste.rs/s/ipfzdh/evolution_rust_compiler_errors#c_wwpcwt][comment at lobste.rs]:

> [There’s a nightly feature which makes the error codes in the output actual links](https://github.com/rust-lang/rust/pull/107838), which kind of sidesteps the issue, but there’s no harm on having a link shortener for this exact use case.

So this might be something that has some similar solution.

**Comment by c-git** [comment]
I'll admit that I didn't appreciate the analogy.

> I'm sorry, this is an accessibility barrier, and that argument is like the owner of a building refusing to add a wheelchair ramp because, "there are still stairs on the second floor." Just because one accessibility barrier might still exist, doesn't justify refusing to fix another major barrier that is even worse.

However my apologies if I let that seep into my response. I was only trying to respond to why and not offend anyone. Just trying to point out my observations based on my tests.

**Comment by lpelecq** [comment]
I encountered the same issue while trying to print a long URL within my test.

In my case I have control over what is printed so I tried OSC 8 terminal hyperlinks but they seem to be stripped out by Cacon (the hyperlink works fine when running `cargo test` directly).

```rs
let esc = "\x1b";
let st = "\x1b\\";
println!("{esc}]8;;{url}{st}Open link in browser{esc}]8;;{st}");
```

Would it be conceivable to make it so that Bacon leaves the escape codes to leave the hyperlinks functionnal ?
